File: src/main/java/de/eugenbernwald/fretboardtrainer/service/PitchEmitter.java
package de.eugenbernwald.fretboardtrainer.service;

import be.tarsos.dsp.AudioDispatcher;
import be.tarsos.dsp.io.jvm.JVMAudioInputStream;
import be.tarsos.dsp.pitch.PitchDetectionHandler;
import be.tarsos.dsp.pitch.PitchProcessor;
import de.eugenbernwald.fretboardtrainer.model.Constants;
import javafx.beans.value.ObservableValue;

import javax.sound.sampled.*;

public class PitchEmitter {

    private Mixer.Info currentMixerInfo;

    private AudioDispatcher dispatcher;

    private AudioFormat audioFormat;
    private PitchDetectionHandler pitchDetectionHandler;

    private PitchProcessor.PitchEstimationAlgorithm currentAlgorithm;

    public PitchEmitter(Mixer.Info mixerInfo, PitchDetectionHandler handler) {
        this.currentAlgorithm = PitchProcessor.PitchEstimationAlgorithm.YIN;
        this.audioFormat = Constants.AUDIO_FORMAT;
        this.currentMixerInfo = mixerInfo;
        this.pitchDetectionHandler = handler;
        this.dispatcher = recreateDispatcher();
    }

    private AudioDispatcher recreateDispatcher() {

        Mixer mixer = AudioSystem.getMixer(currentMixerInfo);
        DataLine.Info dataLineInfo = new DataLine.Info(TargetDataLine.class, audioFormat);

        try {

            TargetDataLine line = (TargetDataLine) mixer.getLine(dataLineInfo);
            line.open(audioFormat, Constants.AUDIO_BUFFER_SIZE);
            line.start();
            AudioInputStream audioInputStream = new AudioInputStream(line);
            JVMAudioInputStream audioStream = new JVMAudioInputStream(audioInputStream);
            this.dispatcher = new AudioDispatcher(audioStream, Constants.AUDIO_BUFFER_SIZE, 0);

        } catch (LineUnavailableException e) {
            e.printStackTrace();
        }

        PitchProcessor processor = new PitchProcessor(
                this.currentAlgorithm,
                Constants.SAMPLE_RATE,
                Constants.AUDIO_BUFFER_SIZE,
                this.pitchDetectionHandler);

        this.dispatcher.addAudioProcessor(processor);

        new Thread(this.dispatcher).start();
        return  this.dispatcher;
    }

    public void stop(){
        this.dispatcher.stop();
    }

    public void onDeviceChange(ObservableValue<? extends Mixer.Info> observable, Mixer.Info oldValue, Mixer.Info newValue) {
        if (this.dispatcher != null) {
            this.dispatcher.stop();
        }
        this.currentMixerInfo = newValue;
        dispatcher = recreateDispatcher();
    }

    public void onAlgorithmChange(
            ObservableValue<? extends PitchProcessor.PitchEstimationAlgorithm> observable,
            PitchProcessor.PitchEstimationAlgorithm oldValue,
            PitchProcessor.PitchEstimationAlgorithm newValue) {

        this.currentAlgorithm = newValue;
        if(this.dispatcher != null){
            this.dispatcher.stop();
        }
        this.dispatcher = recreateDispatcher();
    }

}

File: src/main/java/de/eugenbernwald/fretboardtrainer/service/PitchEmitter.kt
package de.eugenbernwald.fretboardtrainer.service

import be.tarsos.dsp.AudioDispatcher
import be.tarsos.dsp.io.jvm.JVMAudioInputStream
import be.tarsos.dsp.pitch.PitchDetectionHandler
import be.tarsos.dsp.pitch.PitchProcessor
import de.eugenbernwald.fretboardtrainer.model.Constants

import javax.sound.sampled.*

class PitchEmitter(private var currentMixerInfo: Mixer.Info?, private val pitchDetectionHandler: PitchDetectionHandler) {

    private var dispatcher: AudioDispatcher

    private val audioFormat: AudioFormat

    private var currentAlgorithm: PitchProcessor.PitchEstimationAlgorithm

    init {
        currentAlgorithm = PitchProcessor.PitchEstimationAlgorithm.YIN
        audioFormat = Constants.AUDIO_FORMAT
        dispatcher = recreateDispatcher()
    }

    private fun recreateDispatcher(): AudioDispatcher {

        val mixer = AudioSystem.getMixer(currentMixerInfo)
        val dataLineInfo = DataLine.Info(TargetDataLine::class.java, audioFormat)

        try {

            val line = mixer.getLine(dataLineInfo) as TargetDataLine
            line.open(audioFormat, Constants.AUDIO_BUFFER_SIZE)
            line.start()
            val audioInputStream = AudioInputStream(line)
            val audioStream = JVMAudioInputStream(audioInputStream)
            dispatcher = AudioDispatcher(audioStream, Constants.AUDIO_BUFFER_SIZE, 0)

        } catch (e: LineUnavailableException) {
            e.printStackTrace()
        }

        val processor = PitchProcessor(
                currentAlgorithm,
                Constants.SAMPLE_RATE,
                Constants.AUDIO_BUFFER_SIZE,
                pitchDetectionHandler)

        dispatcher.addAudioProcessor(processor)

        Thread(dispatcher).start()
        return dispatcher
    }

    fun stop() {
        dispatcher.stop()
    }

    fun onDeviceChange(newValue: Mixer.Info) {
        dispatcher.stop()
        currentMixerInfo = newValue
        dispatcher = recreateDispatcher()
    }

    fun onAlgorithmChange(newValue: PitchProcessor.PitchEstimationAlgorithm) {

        currentAlgorithm = newValue
        dispatcher.stop()
        dispatcher = recreateDispatcher()
    }

}
